Fix #2476, #3312 and #3703: Move hint handler to domain layer

MaskedCarrot opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

Fixes #2476
Fixes #3312
Fixes #3703

Note that this PR was made in collaboration between @MaskedCarrot, @yashraj-01, and @BenHenning (see commits for specifics). Further, much of this was built as part of #3696 as a separate rewrite to address specific technical needs for both downstream dependent projects (lightweight checkpointing & developer options menu).

At a high-level, this PR moves & rebuilds HintHandler from the UI layer to being a separate, independent logic unit within the domain layer. Key changes that are meant to simplify hint handling:
1. HintHandler was moved to be an interface so that multiple implementations can be provided
2. HelpIndex is now the singular source of truth for all hint/solution state everywhere (domain, checkpoint, and UI); the previous revealed states in Hint/Solution have been removed (which now means those structures are properly immutable for the lifetime of a play session)
3. HintHandler is responsible for everything hints & solution (when there are new ones available, and managing when they've been shown) by being the sole provider of HelpIndex. This includes both computing the available hints/solution as well as managing when hints & the solution have been revealed. This has led to lots of simplifications in the UI.
4. HintHandler's API has been reinvented to be a bit more operation-specific and less generic
5. A monitor was introduced for HintHandler so that the calling controllers know when to re-query for the latest state. There's definitely a data race here, but the data model is eventually consistent so it should be fine.
6. HintHandler's tests make use of lots of new highly specific explorations. These explorations are not accessible from the topic list; they are only included for testing purposes (and are meant to exemplars for future tests).

This PR introduces a few other examples for future changes:
- Module-specific test to verify bindings
- Extensions class tests

While neither of those are unique, the relevant part is that these were the default approaches taken to address the test file presence requirement.

Other details:
- This also fixes #3703 since it was found in passing that //testing module tests were not being built or run in Bazel (except those that are already modularized), leading to some build errors that weren't caught by the CI environment as they should have been.
- QuestionAssessmentProgressControllerTest has been updated to include new scores since it was previously testing impossible scenarios (i.e. assuming hints could be shown/revealed when they wouldn't actually show up)

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

#### Without this fix, tests (added in this PR because they were not present) fail on the latest develop.

![Screenshot from 2021-08-12 03-36-02](https://user-images.githubusercontent.com/53645584/129111400-923f2558-9590-4c0e-8629-3254f13b6abd.png)

Commit:
moved hint handler to domain layer

yashraj-01 committed:
fixed app layer espresso tests

yashraj-01 committed:
fix app layer robolectric tests

yashraj-01 committed:
fixed domain layer tests

yashraj-01 committed:
Added annotations to test exemptions

yashraj-01 committed:
proto lint fix

Commit:
fixed hint handler for training sessions

Commit:
nit and removed test excemptions

Commit:
added hint tests for config change

Commit:
fixed test file exemptions

Commit:
fixed failing test

yashraj-01 committed:
made HintHandler injectable

yashraj-01 committed:
Merge remote-tracking branch 'upstream/move-hint-handler-to-domain' into move-hint-handler-to-domain

yashraj-01 committed:
fixed ktlint error

MaskedCarrot commented:
@BenHenning PTAL

MaskedCarrot commented:
The one failing CI is because there are no tests for the file hintHandler (there are that ensure the proper functioning of the hint handler in state fragment local tests)

yashraj-01 committed:
Added tests for hint handler

MaskedCarrot committed:
Merge branch 'develop' into move-hint-handler-to-domain

Commit:
nits

Commit:
Merge remote-tracking branch 'origin/move-hint-handler-to-domain' into move-hint-handler-to-domain

Commit:
fixed failing build

Commit:
fixed failing build

Commit:
fix build

Commit:
fixed imports

MaskedCarrot committed:
Merge branch 'develop' into move-hint-handler-to-domain

BenHenning commented on app/src/main/java/org/oppia/android/app/player/state/StatePlayerRecyclerViewAssembler.kt:
```diff
@@ -224,9 +205,6 @@ class StatePlayerRecyclerViewAssembler private constructor(
         /* isCorrectAnswer= */ false,
         gcsEntityId
       )
-      if (playerFeatureSet.hintsAndSolutionsSupport) {
```
I feel like we still want to have a feature for this, but maybe to gate whether hints & solutions show up for the user instead of activating the handler.

MaskedCarrot replied:
updated as follows 
```
    if (recyclerViewAssembler.playerHasSupportForHintsAndSolution()) {
      scheduleShowHintAndSolution(ephemeralQuestion.ephemeralState.hintState)
      showHintsAndSolutions(ephemeralQuestion.ephemeralState.hintState)
    }
```

BenHenning commented on model/src/main/proto/exploration.proto:
```diff
@@ -233,9 +233,10 @@ message EphemeralState {
-
   // The current status of checkpointing the exploration.
   CheckpointState checkpoint_state = 7;
+  // The state of hints for the current exploration.
```
Nit: newline before this for spacing.

Ditto below for new protos.

MaskedCarrot replied:
done

BenHenning commented on model/src/main/proto/exploration.proto:
```diff
@@ -335,6 +336,27 @@ message HelpIndex {
 }
 
+// A structure that contains details about the hint that is visible to the learner.
+message HintIndex {
```
Can we consolidate this with HelpIndex to avoid having two different structures?

Note that I specifically called 'HelpIndex' that name so that it could represent both hints & solutions.

MaskedCarrot replied:
removed the message `hintIndex` and updated `helpIndex` as follows
```
message HelpIndex {
  oneof index_type {
    int32 hint_index = 1;
    bool show_solution = 2;
    bool everything_revealed = 3;
}
  bool is_all_visible_help_revealed = 4;
}

```

BenHenning commented on model/src/main/proto/exploration.proto:
```diff
@@ -335,6 +336,27 @@ message HelpIndex {
+  // HelpIndex for the current state.
+  HelpIndex help_index = 1;
+  // Sequence number of the most recently scheduled task to show help to the learner.
+  int32 hint_sequence_number = 2;
+  // Count of wrong answers that the hint handler has been notified about.
+  int32 tracked_answer_count = 3;
+  // Delay for scheduling a new task to show more help to the learner.
+  int64 delay_to_show_next_hint_and_solution = 4;
```
Why do these need to be passed to the UI layer? They all seem like domain-only implementation details (e.g. for the last one, if we don't want a particular hint to show up until after some time then we shouldn't indicate in the ephemeral state that that hint is ready to be viewed by the learner). This may require some minor adjustments to the way we present hint/solution visibility readiness to the UI layer.

MaskedCarrot replied:
The UI layer doesn't need all of these, Right now, it only needs help index, hint_sequence_number and delay_to_show_next_hint_and_solution. These are needed because we have the timer to show help in the UI layer, so to schedule a new task to show help, we need to know the delay and the sequence number (if we move the timer to the domain layer, it will then just need help index)

MaskedCarrot replied:
removed all variables except help_index

BenHenning replied:
Has been resolved in the latest changes.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -204,11 +222,7 @@ class ExplorationProgressController @Inject constructor(
-    hintIsRevealed: Boolean,
-    hintIndex: Int
-  ): LiveData<AsyncResult<Hint>> {
+  fun submitHintIsRevealed(hintIsRevealed: Boolean, hintIndex: Int): LiveData<AsyncResult<Hint>> {
```
While not entirely related to your change, can you please add a KDoc to this method explaining what it does, what parameters are expected, and what it can be expected to return?

MaskedCarrot replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -230,15 +244,28 @@ class ExplorationProgressController @Inject constructor(
+          explorationProgress.stateDeck.pushStateForHint(ephemeralState.state, hintIndex)
         } finally {
+          hintHandler.notifyHintIsRevealed(hintIndex)
+          // Schedule a new hints or solution or show a new hint or solution immediately based on
```
```suggestion
          // Schedule, or show immediately, a new hint or solution based on
```

(May require some re-wrapping with the next line).

MaskedCarrot replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -255,9 +282,7 @@ class ExplorationProgressController @Inject constructor(
-  fun submitSolutionIsRevealed(
-    state: State
-  ): LiveData<AsyncResult<Solution>> {
+  fun submitSolutionIsRevealed(): LiveData<AsyncResult<Solution>> {
```
Ditto for this method (please add a KDoc).

MaskedCarrot replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -397,110 +437,31 @@ class ExplorationProgressController @Inject constructor(
-    lastPlayedTimestamp: Long,
-    newCheckpointState: CheckpointState
-  ) {
+  fun hintAndSolutionTimerCompleted(trackedSequenceNumber: Int, state: State) {
```
I'm wondering if it might be cleaner & simpler to instead send signals to this controller when the timer should be paused/resumed (i.e. during the activity pausing/resuming), but to otherwise perform the timer tracking in the domain layer. My concern with the current implementation is a lot still depends on the UI to do the right things in order for this mechanism to work correctly.

MaskedCarrot replied:
I am not sure how we will move the timer to the domain layer, specifically how we will stop and resume it when the activity resumes or pauses. We can use the override methods onPause and onResume, but how will we notify the timer to pause.

I looked into the default timer that android has,
```
 object : CountDownTimer(dealyToShowHints, 1000) {

     override fun onTick(millisUntilFinished: Long) {}

     override fun onFinish() {
         // reveal hint or solution 
     }
 }.start()
 
```



 Alternatively, we can also use coroutine,
```
    backgroundCoroutineScope.launch {
      delay(delayToShowHints)
      // reveal hint/solution
    }
```

But neither of them have pause and resume methods.

One possible way to handle pause and resume cases is to save the delay left and cancel the timer on pause and rerun a new timer with the remaining delay when the activity is resumed.

```
 object : CountDownTimer(dealyToShowHints, 1000) {

     override fun onTick(millisUntilFinished: Long) {
          if (hintSequenceNumber == mostRecentSequenceNumber) {
               explorationProgress.timeElapsed += 1
         }
     }

     override fun onFinish() {
         // reveal hint or solution 
     }
 }.start()
```
But I think we can do it like this only with the default timers and not with coroutines.

Can you please suggest if the above-mentioned way is the correct way to handle this or give a few points on how I should approach this problem?

MaskedCarrot replied:
Update: 

I think keeping track of the time elapsed (shown in the above block of code in the line `explorationProgress.timeElapsed += 1` ) is not a good way of solving this because then every second lock will have to be acquired to update the elapsed time, this will delay the other functions that are waiting to acquire lock (for eg submit answer)

BenHenning replied:
Hmm. I'm not worried about the locking mechanism (yet) since the functions return relatively quickly.

Regarding how to do this, I think there are a couple ways we can do this that will match the current behavior. However, it occurs to me that the current behavior may be wrong (I hypothesize that if you were to pause the app after triggering the timer & wait for the timer's duration before returning, a hint should show up since the timer continues while the app is backgrounded).

For the sake of time, I suggest we go with a parity approach and file an issue tracking this bug since it'll require more work to solve. Could you please file that issue?

Regarding the solution, I don't think we actually need to do anything extra. The timer will continue while the app is backgrounded, recompute the HelpIndex & EphemeralState, notify subscribers, and post the data to the UI. Fortunately, that last part happens via LiveData in the same way the lifecycle-safe timer does, so the result shouldn't actually be observed until the user returns. I think the current batch of tests should cover these cases well enough that if they pass, the behavior should be working as expected with the timer moved to the domain layer.

MaskedCarrot replied:
moved the timer to the domain layer.

BenHenning replied:
Has been resolved.

BenHenning commented on app/src/test/java/org/oppia/android/app/player/state/StateFragmentLocalTest.kt:
```diff
@@ -762,6 +762,45 @@ class StateFragmentLocalTest {
   }
 
+  @Test
+  fun testStateFragment_nextState_viewHint_configChange_noNewHintAvailable() {
```
It's great to see these tests added. Can you briefly verify that these fail without the other fixes in place & add screenshots to your main PR comment showing these tests passing/failing based on the fix? It's useful to do that for bug fixes to be really clear that the tests will prevent the behavior from regressing.

MaskedCarrot replied:
done

BenHenning commented on app/src/test/java/org/oppia/android/app/player/state/StateFragmentLocalTest.kt:
```diff
@@ -762,6 +762,45 @@ class StateFragmentLocalTest {
+      produceAndViewFirstHint()
+
+      submitWrongAnswerToState2()
+      onView(isRoot()).perform(orientationLandscape())
```
I'm actually wondering whether this will correctly repro the issue. I believe @FareesHussain had another way to solve the problem but couldn't actually write tests for it due to Robolectric limitations. Curious to see per the above if these tests actually pass/fail as expected.

MaskedCarrot replied:
This approach solves the two existing issues that were present in the hint hander
1. Hints that were scheduled to be visible when the timer runs out, are visible even on the completed state.
2. new hints were revealed (after the first hint was viewed) upon configuration change.

I have verified that without this fix the tests that check the above two cases fail (added screenshot in pr description).

All tests pass for the current approach.

BenHenning replied:
Interesting. Thanks for checking @MaskedCarrot. I'm surprised at the discrepancy in how Robolectric is handling these config changes, but I'm glad to see the tests are correctly covering the failure case.

BenHenning commented on app/src/test/java/org/oppia/android/app/player/state/StateFragmentLocalTest.kt:
```diff
@@ -762,6 +762,45 @@ class StateFragmentLocalTest {
+  }
+
+  @Test
+  fun testStateFragment_nextState_viewHint_prevState_wait30seconds_newHintIsAvailable() {
```
Should this read as not available?

MaskedCarrot replied:
thanks, I must have missed that

BenHenning commented on app/src/test/java/org/oppia/android/app/player/state/StateFragmentLocalTest.kt:
```diff
@@ -762,6 +762,45 @@ class StateFragmentLocalTest {
+      submitWrongAnswerToState2()
+      onView(isRoot()).perform(orientationLandscape())
+      testCoroutineDispatchers.runCurrent()
+      onView(withId(R.id.dot_hint)).check(matches(not(isDisplayed())))
```
Should the test name indicate that a hint is actually not available given this?

MaskedCarrot replied:
No, this test indicates that if the user views a hint and then changes the device configuration, the next hint should not be available. Updated the function name to clearly indicate this.

BenHenning replied:
Much clearer--thanks.

BenHenning reviewed (changes requested):
Thanks @MaskedCarrot. I think this is a really awesome improvement. Left an initial round of comments--will take a full pass after these are addressed. I looked at mostly everything, but I didn't look at the new test suite in detail (yet).

Note that after these initial comments are resolved, please make sure that all public API method changes are covered in tests as well (as I'll likely request them to be added in my next review, anyway).

Commit:
nits and improved testing

Commit:
updated exploration.proto

Commit:
removed progress controller from kdoc exemptions

Commit:
Merge remote-tracking branch 'origin/develop' into move-hint-handler-to-domain

Commit:
fix failing test

BenHenning commented on app/src/main/java/org/oppia/android/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -323,8 +323,11 @@ class StateFragmentPresenter @Inject constructor(
+
+    if (recyclerViewAssembler.playerHasSupportForHintsAndSolution()) {
+      scheduleShowHintAndSolution(ephemeralState.hintState)
+      showHintsAndSolutions(ephemeralState.hintState)
```
Why not call these as part of the assembler's compute method? It's a bit odd/unexpected to be implementing assembler feature behavior checking outside of the assembler, and it seems possible to fold this directly into the assembler itself (& also gain the benefit of sharing code between the exploration & question player as is the purpose of the assembler).

MaskedCarrot replied:
having the listener works, updated the PR.

BenHenning replied:
Has been resolved.

BenHenning commented on app/src/main/java/org/oppia/android/app/player/state/StatePlayerRecyclerViewAssembler.kt:
```diff
@@ -273,6 +273,8 @@ class StatePlayerRecyclerViewAssembler private constructor(
     return Pair(conversationPendingItemList, extraInteractionPendingItemList)
   }
 
+  fun playerHasSupportForHintsAndSolution(): Boolean = playerFeatureSet.hintsAndSolutionsSupport
```
Can be omitted, I think, per my other comment.

MaskedCarrot replied:
removed.

BenHenning replied:
Has been resolved.

BenHenning commented on model/src/main/proto/exploration.proto:
```diff
@@ -334,25 +350,23 @@ message HelpIndex {
-  // Boolean to indicate if the learner has revealed the hint or not.
-  bool is_hint_revealed = 2;
+  // Boolean to indicate if the learner has revealed the latest  help or not.
+  bool is_all_visible_help_revealed = 4;
```
Nit: double space between 'latest' and 'help'.

Also, the name & the description seem at odds. The doc comment seems to indicate whether the latest one was revealed, but the name suggests "everything" being revealed. I think you meant the former since we have 'everything_revealed' above for the latter. Can you rename this for clarity?

Also, I'm wondering if we could bake this into the oneof somehow. Maybe:

```proto
message HelpIndex {
  oneof index_type {
    int32 available_next_hint_index = 1;
    int32 latest_revealed_hint_index = 2;
    bool show_solution = 3;
    bool everything_revealed = 4;
  }
}
```

Where the default state (unspecified) is "nothing is ready yet" and each of the states implies all previous hints are available. What do you think about this alternative structure?

MaskedCarrot replied:
this will work. Thanks, changed the structure.

BenHenning replied:
Has been resolved.

BenHenning commented on scripts/assets/kdoc_validity_exemptions.textproto:
```diff
@@ -376,7 +376,6 @@ exempted_file_path: "domain/src/main/java/org/oppia/android/domain/classify/rule
 exempted_file_path: "domain/src/main/java/org/oppia/android/domain/classify/rules/InteractionObjectTypeExtractorRepository.kt"
 exempted_file_path: "domain/src/main/java/org/oppia/android/domain/classify/rules/continueinteraction/ContinueModule.kt"
 exempted_file_path: "domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgress.kt"
-exempted_file_path: "domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt"
```
Awesome! Glad to see one more file removed from this list. :)

Thanks for filling in the missing KDocs.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentProgressController.kt:
```diff
@@ -245,6 +255,12 @@ class QuestionAssessmentProgressController @Inject constructor(
   }
 
+  /**
+   *  Notifies the [StateDeck] and the [HintHandler] that the solution has been revealed by the user.
```
```suggestion
   * Notifies the [StateDeck] and the [HintHandler] that the solution has been revealed by the user.
```

MaskedCarrot replied:
done

BenHenning replied:
Has been resolved.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/hintsandsolution/HintHandler.kt:
```diff
@@ -109,14 +108,16 @@ class HintHandler @Inject constructor(
           hintSequenceNumber++
           delayShowInitialHintMs
         } else {
+          // Update the sequence number to cancel any pending tasks ans show new help immediately.
```
```suggestion
          // Update the sequence number to cancel any pending tasks and show new help immediately.
```

MaskedCarrot replied:
removed this comment because removed the line below it.

BenHenning replied:
Has been resolved.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/hintsandsolution/HintHandler.kt:
```diff
@@ -109,14 +108,16 @@ class HintHandler @Inject constructor(
           delayShowInitialHintMs
         } else {
+          // Update the sequence number to cancel any pending tasks ans show new help immediately.
+          hintSequenceNumber++
```
This seems...unexpected. We are now unconditionally always incrementing the sequence number for this index type--is that actually correct? I wouldn't expect a behavior change since we're just moving this to the domain layer.

MaskedCarrot replied:
we don't need this because no new help will be scheduled until a delay other than -1 is available. Removed this.

BenHenning replied:
Has been resolved.

BenHenning reviewed (commented):
Thanks @MaskedCarrot. Followed up on existing discussions.

MaskedCarrot committed:
moved timer to domain

MaskedCarrot committed:
fixed build and nits

MaskedCarrot committed:
nit

MaskedCarrot committed:
added listener back to test exemptions

vinitamurthi commented on domain/src/main/java/org/oppia/android/domain/hintsandsolution/HintHandler.kt:
```diff
@@ -0,0 +1,210 @@
+package org.oppia.android.domain.hintsandsolution
```
Just want to clarify -- this is an exact copy of the existing hint handler correct?

MaskedCarrot replied:
No this is not an exact copy of the previous hint handler. 

The previous hint handler was responsible for the following
- compute which hint (if any) should be revealed to the user.
- schedule a new timer, observe it and show new hints when the timer runs out.

The new hint handler computes which hint should be revealed but it is no longer responsible to schedule timer to show help, instead, the hint handler computes the delay after which the hint should be revealed. The timer is scheduled by the exploration progress controller using the delay computed by the hint handler.

vinitamurthi commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -180,11 +191,29 @@ class ExplorationProgressController @Inject constructor(
+            // Reset the hintState if pending top state has changed.
+            explorationProgress.hintState = hintHandler.reset()
+          } else {
+            //  Schedule, or show immediately, a new hint or solution based on the current ephemeral
```
Dont we need test cases in ExplorationProgressControllerTest for this scenario?

MaskedCarrot replied:
added more tests to test this.

BenHenning replied:
Can this now be marked as resolved?

BenHenning replied:
This appears to be resolved (tests in ExplorationProgressControllerTest include cases for wrong answer submission--I verified by removing the new version of this code & observed 9 test failures in ExplorationProgressControllerTest). Closing the thread with permission from @vinitamurthi to verify & close out old conversation threads.

vinitamurthi commented on domain/src/test/java/org/oppia/android/domain/hintsandsolution/HintHandlerTest.kt:
```diff
@@ -0,0 +1,468 @@
+package org.oppia.android.domain.hintsandsolution
```
Again, to clarify -- this is an exact copy of the app module test?

MaskedCarrot replied:
We did not have these tests previously, this is a new file.

BenHenning replied:
Can this be marked as resolved?

BenHenning replied:
This appears to be resolved (the test suite is entirely new, and has been vetted/written by me). Closing the thread with permission from @vinitamurthi to verify & close out old conversation threads.

vinitamurthi commented on domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentProgressController.kt:
```diff
@@ -157,13 +171,31 @@ class QuestionAssessmentProgressController @Inject constructor(
               )
             }
+          } else {
+            // Schedule a new hints or solution or show a new hint or solution immediately based on
```
Do we need tests for this scenario in QuestionAssessmentProgressControllerTest?

MaskedCarrot replied:
We have another issue tracking this (#1273). The state fragment also uses the same logic for hints and solutions and we do have tests in state fragment local tests that test this behaviour.

BenHenning replied:
Similarly, HintHandlerImplTest is thoroughly being tested and ensures consistent behavior between questions & explorations. Unfortunately, we have only surface-level checks to ensure that the fully questions stack is properly integrating with these systems. We could definitely improve testing coverage for the latter in the future.

vinitamurthi commented on model/src/main/proto/exploration.proto:
```diff
@@ -319,22 +336,40 @@ message HelpIndex {
-    int32 hint_index = 1;
+    // Indicates this help index corresponds to the index of the next available hint within the hint
+    // list of a state.
+    int32 available_next_hint_index = 1;
```
In general we shouldn't do modifications to protos like this (see more here https://developers.google.com/protocol-buffers/docs/proto3#updating) . Can we instead deprecate hint_index and add the new fields in the end? i.e.
```
reserved 1;

// ...
bool show_solution = 2;

// ...
bool everything_revealed = 3;

// ...
int32 available_next_hint_index = 4;

// ...
int32 latest_revealed_hint_index = 5;
```

MaskedCarrot replied:
done

BenHenning replied:
FWIW this is an ephemeral message so breaking compatibility is fine, but it's a good suggestion in principle.

Also, can this now be marked as resolved?

BenHenning replied:
This appears to be resolved (mainly since the old index has now been reserved). Closing the thread with permission from @vinitamurthi to verify & close out old conversation threads.

vinitamurthi commented on model/src/main/proto/exploration.proto:
```diff
@@ -319,22 +336,40 @@ message HelpIndex {
+  int32 hint_sequence_number = 2;
+
+  // Delay for scheduling a new task to show more help to the learner.
+  int64 delay_to_show_next_hint_and_solution = 4;
```
```suggestion
  int64 delay_to_show_next_hint_and_solution = 3;
```

MaskedCarrot replied:
done

BenHenning replied:
Can this be resolved now?

BenHenning replied:
This appears to be resolved (mainly since HintState no longer exists). Closing the thread with permission from @vinitamurthi to verify & close out old conversation threads.

vinitamurthi commented on model/src/main/proto/exploration_checkpoint.proto:
```diff
@@ -46,6 +46,8 @@ message ExplorationCheckpoint {
   // The timestamp in milliseconds of when the checkpoint was saved for the first time.
   int64 timestamp_of_first_checkpoint = 9;
+
+  HelpIndex help_index = 10;
```
Add a comment explaining the field

MaskedCarrot replied:
done

MaskedCarrot committed:
nit fixes and added more tests

MaskedCarrot committed:
lint fix

rt4914 commented on app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerFragmentPresenter.kt:
```diff
@@ -51,7 +51,7 @@ class QuestionPlayerFragmentPresenter @Inject constructor(
   @QuestionResourceBucketName private val resourceBucketName: String,
   private val assemblerBuilderFactory: StatePlayerRecyclerViewAssembler.Builder.Factory,
-  private val splitScreenManager: SplitScreenManager
+  private val splitScreenManager: SplitScreenManager,
```
Why is there `,` at the end?

rt4914 reviewed (approved):
Approving for code owner files. Just left one comment. Will mainly defer to @vinitamurthi and @BenHenning for approval.

BenHenning commented on model/src/main/proto/exploration.proto:
```diff
@@ -332,9 +350,30 @@ message HelpIndex {
+
+    // Indicates this help index corresponds to the index of the next available hint within the hint
+    // list of a state.
+    int32 available_next_hint_index = 4;
```
```suggestion
    int32 next_available_hint_index = 4;
```

Ditto for the comment above.

BenHenning replied:
Addressed in the last changes.

BenHenning commented on model/src/main/proto/exploration.proto:
```diff
@@ -332,9 +350,30 @@ message HelpIndex {
+    // list of a state.
+    int32 available_next_hint_index = 4;
+
+    // Indicates this help index corresponds to the index of the last revealed hint within the hint
+    // list of a state.
```
I think we're missing some important context in this & the above doc comment. The case above specifically indicates that a new hint is available to be revealed whereas this case represents that all available hints have been revealed. Without that distinction these cases seem like they could be combined into one.

BenHenning replied:
Done (in the latest changes).

BenHenning commented on model/src/main/proto/exploration.proto:
```diff
@@ -332,9 +350,30 @@ message HelpIndex {
 
+// A Structure corresponding to the state of hint and solution for the current exploration or
+// training session.
+message HintState {
+  // HelpIndex for the current state.
+  HelpIndex help_index = 1;
+
+  // Sequence number of the most recently scheduled task to show help to the learner.
+  int32 hint_sequence_number = 2;
+
+  // Delay for scheduling a new task to show more help to the learner.
+  int64 delay_to_show_next_hint_and_solution = 3;
+}
```
I thought per our discussion only HelpIndex needs to be exposed--why is this needed?

MaskedCarrot replied:
only help index is being exposed to the UI, these values are being stored in exploration Progress. we need to have these in the progress controller to correctly handler the timer that shows new help

BenHenning replied:
No longer needed (I think); HelpIndex is now sufficient due to a bunch of pipelining changes. This is now resolved.

BenHenning commented on model/src/main/proto/exploration_checkpoint.proto:
```diff
@@ -46,6 +46,9 @@ message ExplorationCheckpoint {
   // The timestamp in milliseconds of when the checkpoint was saved for the first time.
   int64 timestamp_of_first_checkpoint = 9;
+
+  // The saved help index for the exploration.
```
```suggestion
  // The saved help index for the exploration indicating the revealed & available hints/solution of the latest state.
```

BenHenning replied:
Also, are 'solution_is_revealed' & 'hint_index' still needed now that this is here?

BenHenning replied:
Those fields have been removed since only HelpIndex is needed now.

BenHenning reviewed (commented):
Sending some initial comments--I might actually follow up with direct code changes here.

BenHenning committed:
First round: make HintHandler independent.

This also brings HintHandler into an interface + factory pattern. This
isn't the final design since I think we can largely simplify the way
hints work; that'll be my next pass.

This breaks questions & HintHandlerTest; those will require further work
later.

vinitamurthi reviewed (approved):
LGTM for codeowner files

BenHenning committed:
Simplify HelpIndex in proto.

Move HelpIndex to PendingState to avoid the entire domain case of
handling CompletedState.

BenHenning committed:
Simplify hints & solutions.

This removes the per-Hint/Solution tracking & completely leans on
HelpIndex for proper hints & solution tracking both in the domain & UI
layers.

Fixes a bunch of other stuff, too, including
QuestionAssessmentProgressController tests.

BenHenning committed:
Clean up dead code paths & improve handler API.

yashraj-01 committed:
fixed test modules and lint

yashraj-01 committed:
renamed HintHandlerTest to HintHandlerImplTest

BenHenning committed:
Add tests for HintHandler.

This introduces some new explorations for making testing HintHandlerImpl
easier.

BenHenning committed:
Add remaining tests/exemptions for new files.

BenHenning committed:
Lint fixes.

BenHenning committed:
Merge branch 'develop' into move-hint-handler-to-domain

Conflicts:
	app/src/main/java/org/oppia/android/app/application/ApplicationComponent.kt
	app/src/sharedTest/java/org/oppia/android/app/home/TopicSummaryViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/WelcomeViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/promotedlist/PromotedStoryListViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/promotedlist/PromotedStoryViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/parser/CustomBulletSpanTest.kt
	app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
	app/src/sharedTest/java/org/oppia/android/app/splash/SplashActivityTest.kt
	app/src/test/java/org/oppia/android/app/home/HomeActivityLocalTest.kt
	app/src/test/java/org/oppia/android/app/player/exploration/ExplorationActivityLocalTest.kt
	app/src/test/java/org/oppia/android/app/profile/ProfileChooserFragmentLocalTest.kt
	app/src/test/java/org/oppia/android/app/testing/CompletedStoryListSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/HomeSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/OngoingTopicListSpanTest.kt
app/src/test/java/org/oppia/android/app/testing/ProfileChooserSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/ProfileProgressSpanCount.kt
	app/src/test/java/org/oppia/android/app/testing/RecentlyPlayedSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/TopicRevisionSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/player/split/PlayerSplitScreenTesting.kt
	app/src/test/java/org/oppia/android/app/testing/player/state/StateFragmentAccessibilityTest.kt
	app/src/test/java/org/oppia/android/app/topic/revisioncard/RevisionCardActivityLocalTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/question/QuestionTrainingControllerTest.kt

BenHenning committed:
Post-merge fixes (including lint fixes).

BenHenning committed:
Merge branch 'develop' into move-hint-handler-to-domain

Conflicts:
	third_party/maven_install.json

BenHenning committed:
Post-merge maven_install fix.

anandwana001 commented on domain/src/test/java/org/oppia/android/domain/hintsandsolution/HintHandlerTest.kt:
```diff
@@ -0,0 +1,468 @@
+      HelpIndex.IndexTypeCase.AVAILABLE_NEXT_HINT_INDEX
+    )
+    assertThat(currentHintState.helpIndex.availableNextHintIndex).isEqualTo(0)
+    // The delay should now be equal to -1 because there is unrevealed help available.
```
```suggestion
    // The delay should now be equal to -1 because there is unrevealed hint available.
```

BenHenning replied:
No longer relevant since the suite has been rewritten. Can this be marked as resolved now?

anandwana001 commented on domain/src/test/java/org/oppia/android/domain/hintsandsolution/HintHandlerTest.kt:
```diff
@@ -0,0 +1,468 @@
+
+  /**
+   * Creates a blank subscription to the current state to ensure that requests to load the
+   * exploration complete, otherwise post-load operations may fail. An observer is required since
```
```suggestion
   * exploration is complete, otherwise, post-load operations may fail. An observer is required since
```

BenHenning replied:
Ditto--suite has been rewritten. Can this be marked as resolved now?

anandwana001 reviewed (approved):
LGTM for code owner files,

nit changes suggested below. Fix the merge conflicts.

MaskedCarrot committed:
Merge branch 'develop' into move-hint-handler-to-domain

MaskedCarrot committed:
Revert "Merge branch 'develop' into move-hint-handler-to-domain"

This reverts commit e2fea908a65af8c11b5ed135f59cec38e807f251, reversing
changes made to 6659858f0b85a0bf0c51f23af600769eac64b938.

BenHenning committed:
Post-merge Gradle-discovered fixes.

BenHenning committed:
Revert "Revert "Merge branch 'develop' into move-hint-handler-to-domain""

This reverts commit b1622c00d33f584ae727f0c9c4434ad773bc80a2.

BenHenning committed:
Merge branch 'attempt-to-fix-move-hint-handler-to-domain' into move-hint-handler-to-domain

Conflicts:
	app/src/sharedTest/java/org/oppia/android/app/home/TopicSummaryViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/WelcomeViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/promotedlist/PromotedStoryListViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/promotedlist/PromotedStoryViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/parser/CustomBulletSpanTest.kt
	app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
	app/src/sharedTest/java/org/oppia/android/app/splash/SplashActivityTest.kt
	app/src/test/java/org/oppia/android/app/home/HomeActivityLocalTest.kt
	app/src/test/java/org/oppia/android/app/player/exploration/ExplorationActivityLocalTest.kt
	app/src/test/java/org/oppia/android/app/profile/ProfileChooserFragmentLocalTest.kt
	app/src/test/java/org/oppia/android/app/testing/CompletedStoryListSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/HomeSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/OngoingTopicListSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/ProfileChooserSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/ProfileProgressSpanCount.kt
	app/src/test/java/org/oppia/android/app/testing/RecentlyPlayedSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/TopicRevisionSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/player/split/PlayerSplitScreenTesting.kt
	app/src/test/java/org/oppia/android/app/testing/player/state/StateFragmentAccessibilityTest.kt
	app/src/test/java/org/oppia/android/app/topic/revisioncard/RevisionCardActivityLocalTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/question/QuestionTrainingControllerTest.kt

BenHenning commented:
I've verified that the incoming changes build & pass on Bazel. I verified that a subset of them pass all CI checks & Gradle, though with recent merging & other changes it's possible some of these have regressed.

BenHenning committed:
Additional post-merge fixes.

BenHenning commented:
@rt4914 @anandwana001 @vinitamurthi unfortunately, I think you will each need to take one more pass on this PR. It's been almost entirely rewritten, and mostly by me which means I shouldn't be the one to approve it for merge. I'll still take a pass to double check my work & ensure the bits of code I didn't write mix correctly with the bits I did.

PTAL with priority since this blocks two GSoC projects. Note that this solution has been heavily vetted and I'm confident we can proceed with it; we just need the review completed at this point.

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkStoriesCompletedActivityTest.kt:
```diff
@@ -157,13 +158,13 @@ class MarkStoriesCompletedActivityTest {
       DeveloperOptionsStarterModule::class, DeveloperOptionsModule::class,
       ExplorationStorageModule::class, NetworkConnectionUtilDebugModule::class,
-      NetworkConnectionDebugUtilModule::class
+      NetworkConnectionDebugUtilModule::class,
```
nit fix 
```suggestion
      NetworkConnectionDebugUtilModule::class
```

anandwana001 replied:
All other places as well.

rt4914 replied:
+1

BenHenning replied:
Done.

anandwana001 reviewed (approved):
Test files LGTM so far, keeping eye on checks for the result of all the running tests.

vinitamurthi commented on domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentProgress.kt:
```diff
@@ -24,7 +24,9 @@ internal class QuestionAssessmentProgress {
 
   /** Initialize the assessment with the specified list of questions. */
-  internal fun initialize(questionsList: List<Question>) {
+  internal fun initialize(
+    questionsList: List<Question>
+  ) {
```
Is this wrapping needed?

BenHenning replied:
Nope; not sure where this happened but fixed.

vinitamurthi reviewed (approved):
LGTM. Note that the approval is for domain layer and the proto files (since thats what I am a codeowner for)

rt4914 commented on app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerFragmentPresenter.kt:
```diff
@@ -51,7 +49,7 @@ class QuestionPlayerFragmentPresenter @Inject constructor(
   @QuestionResourceBucketName private val resourceBucketName: String,
   private val assemblerBuilderFactory: StatePlayerRecyclerViewAssembler.Builder.Factory,
-  private val splitScreenManager: SplitScreenManager
+  private val splitScreenManager: SplitScreenManager,
```
Remove `,` at the end.

BenHenning replied:
Done.

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/administratorcontrols/AppVersionActivityTest.kt:
```diff
@@ -264,13 +265,13 @@ class AppVersionActivityTest {
       DeveloperOptionsStarterModule::class, DeveloperOptionsModule::class,
       ExplorationStorageModule::class, NetworkConnectionUtilDebugModule::class,
-      NetworkConnectionDebugUtilModule::class
+      NetworkConnectionDebugUtilModule::class,
```
Remove `,` at the end

BenHenning replied:
Done.

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/completedstorylist/CompletedStoryListActivityTest.kt:
```diff
@@ -484,13 +485,13 @@ class CompletedStoryListActivityTest {
       DeveloperOptionsStarterModule::class, DeveloperOptionsModule::class,
       ExplorationStorageModule::class, NetworkConnectionUtilDebugModule::class,
-      NetworkConnectionDebugUtilModule::class
+      NetworkConnectionDebugUtilModule::class,
```
Ditto

BenHenning replied:
Done.

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/customview/LessonThumbnailImageViewTest.kt:
```diff
@@ -142,13 +143,13 @@ class LessonThumbnailImageViewTest {
       DeveloperOptionsStarterModule::class, DeveloperOptionsModule::class,
       ExplorationStorageModule::class, NetworkConnectionUtilDebugModule::class,
-      NetworkConnectionDebugUtilModule::class
+      NetworkConnectionDebugUtilModule::class,
```
Ditto

BenHenning replied:
Done.

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/databinding/MarginBindingAdaptersTest.kt:
```diff
@@ -280,13 +281,13 @@ class MarginBindingAdaptersTest {
       DeveloperOptionsStarterModule::class, DeveloperOptionsModule::class,
       ExplorationStorageModule::class, NetworkConnectionUtilDebugModule::class,
-      NetworkConnectionDebugUtilModule::class
+      NetworkConnectionDebugUtilModule::class,
```
Ditto

BenHenning replied:
Done.

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerMarginBindingAdaptersTest.kt:
```diff
@@ -468,13 +469,13 @@ class StateAssemblerMarginBindingAdaptersTest {
       DeveloperOptionsStarterModule::class, DeveloperOptionsModule::class,
       ExplorationStorageModule::class, NetworkConnectionUtilDebugModule::class,
-      NetworkConnectionDebugUtilModule::class
+      NetworkConnectionDebugUtilModule::class,
```
Ditto

BenHenning replied:
Done.

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/databinding/StateAssemblerPaddingBindingAdaptersTest.kt:
```diff
@@ -466,13 +467,13 @@ class StateAssemblerPaddingBindingAdaptersTest {
       DeveloperOptionsStarterModule::class, DeveloperOptionsModule::class,
       ExplorationStorageModule::class, NetworkConnectionUtilDebugModule::class,
-      NetworkConnectionDebugUtilModule::class
+      NetworkConnectionDebugUtilModule::class,
```
Ditto

BenHenning replied:
Done.

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/databinding/ViewBindingAdaptersTest.kt:
```diff
@@ -200,13 +201,13 @@ class ViewBindingAdaptersTest {
       DeveloperOptionsStarterModule::class, DeveloperOptionsModule::class,
       ExplorationStorageModule::class, NetworkConnectionUtilDebugModule::class,
-      NetworkConnectionDebugUtilModule::class
+      NetworkConnectionDebugUtilModule::class,
```
Ditto

BenHenning replied:
Done.

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsActivityTest.kt:
```diff
@@ -268,13 +268,13 @@ class DeveloperOptionsActivityTest {
       DeveloperOptionsStarterModule::class, DeveloperOptionsModule::class,
       ExplorationStorageModule::class, NetworkConnectionUtilDebugModule::class,
-      NetworkConnectionDebugUtilModule::class
+      NetworkConnectionDebugUtilModule::class,
```
Ditto

BenHenning replied:
Done.

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/devoptions/DeveloperOptionsFragmentTest.kt:
```diff
@@ -518,13 +519,13 @@ class DeveloperOptionsFragmentTest {
       DeveloperOptionsStarterModule::class, DeveloperOptionsModule::class,
       ExplorationStorageModule::class, NetworkConnectionUtilDebugModule::class,
-      NetworkConnectionDebugUtilModule::class
+      NetworkConnectionDebugUtilModule::class,
```
Ditto

BenHenning replied:
Done.

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedActivityTest.kt:
```diff
@@ -157,13 +158,13 @@ class MarkChaptersCompletedActivityTest {
       DeveloperOptionsStarterModule::class, DeveloperOptionsModule::class,
       ExplorationStorageModule::class, NetworkConnectionUtilDebugModule::class,
-      NetworkConnectionDebugUtilModule::class
+      NetworkConnectionDebugUtilModule::class,
```
Ditto

BenHenning replied:
Done.

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/devoptions/MarkChaptersCompletedFragmentTest.kt:
```diff
@@ -861,13 +862,13 @@ class MarkChaptersCompletedFragmentTest {
       DeveloperOptionsStarterModule::class, DeveloperOptionsModule::class,
       ExplorationStorageModule::class, NetworkConnectionUtilDebugModule::class,
-      NetworkConnectionDebugUtilModule::class
+      NetworkConnectionDebugUtilModule::class,
```
Ditto

BenHenning replied:
Done.

rt4914 reviewed (approved):
LGTM. There is one consistent change across multiple files: there is one unnecessary `,` at the end. Please remove that from all files.

BenHenning committed:
Address first batch of review comments.

(Clarified some proto fields & remove trailing comma).

BenHenning commented:
Note that this will now also fix #3703 which was found in passing while addressing review comments.

BenHenning committed:
Fix testing module tests & remove unnecessary changes.

BenHenning committed:
Simplify & fix more reviewer comments.

This simplifies some data pipelining in the UI, and improves HintHandler
documentation in addition to fixing some more reviewer comments.

yashraj-01 commented:
linking issue #3312 as this PR essentially covers the scope of the issue.

BenHenning committed:
Improve documentation to address review comment.

BenHenning committed:
Rename new module to prod module.

This simplifies the changes & approvals needed in #3705.

BenHenning committed:
Rename proto field (to address review comment).

Also, fix broken reference error accidentally introduced in an earlier
commit.

BenHenning committed:
Kotlin lint fixes.

BenHenning committed:
Merge branch 'develop' into move-hint-handler-to-domain

BenHenning reviewed (approved):
Thanks for the fast reviews! @rt4914 @vinitamurthi @anandwana001 can you PTAL at the latest changes & make sure they look good (& that the old conversation threads are resolved if so)?

I verified that all static checks, Gradle tests, and Bazel tests seem to work locally before uploading the latest changes. I also synced the branch with develop. If no further questions arise, this PR can probably be merged without waiting for the Bazel tests (to speed things up for @yashraj-01 & @MaskedCarrot).

Also, marking PR as approved since I reviewed all the changes & verified my old comment threads are now addressed.

BenHenning commented:
Ack @yashraj-01 that SGTM. Thanks!

BenHenning commented:
Ah excellent. We can see via CI that #3703 is being resolved (all //testing module tests are now showing up in CI as they should have been originally).

anandwana001 reviewed (approved):
LGTM,

as per [this](https://github.com/oppia/oppia-android/pull/3659#pullrequestreview-733406199) comment all static and bazel test checked locally, so the PR is gtg.

rt4914 reviewed (approved):
LGTM

BenHenning commented:
All conversation threads have been verified & closed, required CI checks are passing, and required approvals are in. Going ahead and merging this since I verified that the new testing tests are correctly in CI now, and I verified locally that all Bazel tests are passing. This is blocking two separate streams of time-sensitive work, so I think we can take the risk of merging this in now.

Note that it's theoretically possible that the testing module tests that were previously missing from CI now fail on develop. If this occurs, I'll fix forward to avoid reverting this large, blocking PR & because those tests are not yet blocking (we're running them in Gradle & they've been passing on CI--that check _is_ required).

Thanks again all for the repeated reviews & collaboration on this tricky unexpected design challenge. I'm glad we found a workable solution quickly.
